### DISCUSSION TITLE
New Compilation

### DISCUSSION THREAD

**Discussion by biqas** [dotnet/roslyn#75439] [Q&A] [answered]
Hello,

in previous Version 4.8.0 getting a compilation from a project and then removing all syntaxtrees from that compilation and adding a new syntaxtree to the same compilation (tree has no errors) has semantic errors now, before it worked. Errors are: system references are missing (like Func<>, Span<>, ...). Checking the compilation for errors before changing the trees everything is fine?!?

**Comment by 333fred** [comment]
Can you please be more specific? Provide some sample code, what you expect the behavior to be, and what it is instead?

**Reply by biqas** [reply]
Here some code:

```csharp
var originCompilation = await originProject.GetCompilation(token);

var templateCompilation = originCompilation
    .RemoveAllSyntaxTrees()
    .AddSyntaxTrees(template.SyntaxTree); // Syntax Tree Has No Errors, Is Checked.

var m = originCompilation.GetSemanticModel(originCompilation.SyntaxTrees.First());
this.HasSemanticErrors(m, token); // No Errors => Compilation Is Fine.

m = templateCompilation.GetSemanticModel(templateCompilation.SyntaxTrees.First());

this.HasSemanticErrors(m, token); // Errors: Func<> Not Found For Example.
```

the second

`
this.HasSemanticErrors(m, token);
`
has errors, but there should be not that kind of errors, that core references are missing now, i just swapped the trees not the references.

**Reply by 333fred** [reply]
What is `HasSemanticErrors`? How is `originProject` created? Can you provide a complete code example?

**Reply by jasonmalinowski** [reply]
__Putting @biqas's response to @333fred in a comment here:__

solution comes from MSBuildWorkspace and the projects in the solution are valid projects.

`originProject = solution.Projects.First();`

`
HasSemanticErrors => semanticModel
            .GetDiagnostics(cancellationToken: token)
            .Where(x => x.Severity == DiagnosticSeverity.Error)
            .Select(x => x.ToString())
            .ToArray().Length > 0;
`

and template.SyntaxTree is something like this:
```
using System;

Console.WriteLine();
```
The example is simplified so it is not exact, but the relevant parts are there.

**Reply by jasonmalinowski** [reply]
> The example is simplified so it is not exact, but the relevant parts are there.

The details often matter in the world of compilers! One thing that comes to mind -- if one of the trees you removed was the tree that contained global usings, that could impact other files.

**Reply by biqas** [reply, marked as answer]
I guess this was in this instance the detail which mattered :), tried to just add the template.SyntaxTree and it has no Errors anymore. Thx.
